tensorflow/lite/micro/kernels/xtensa_hifi/depthwise_conv.cc

```diff
@@ -0,0 +1,597 @@
+    const int stride_height = params->stride_height;
+    const int dilation_width_factor = 1 ;
+    const int dilation_height_factor = 1;
+    //const int dilation_width_factor = params->dilation_width_factor;
+    //const int dilation_height_factor = params->dilation_height_factor;
+    const int pad_width = data->padding.width;
+    const int pad_height = data->padding.height;
+    const int depth_multiplier = params->depth_multiplier;
+    const int32 output_activation_min = data->output_activation_min;
+    const int32 output_activation_max = data->output_activation_max;
+    const int32 output_multiplier = data->output_multiplier;
+    // Legacy ops used mixed left and right shifts. Now all are +ve-means-left.
+    const int output_shift = -data->output_shift;
+    TFLITE_DCHECK_EQ(input_shape.DimensionsCount(), 4);
+    TFLITE_DCHECK_EQ(filter_shape.DimensionsCount(), 4);
+    TFLITE_DCHECK_EQ(output_shape.DimensionsCount(), 4);
+
+    TFLITE_DCHECK_LE(output_activation_min, output_activation_max);
+    const int batches = MatchingDim(input_shape, 0, output_shape, 0);
+    const int output_depth = MatchingDim(filter_shape, 3, output_shape, 3);
+    const int input_height = input_shape.Dims(1);
+    const int input_width = input_shape.Dims(2);
+    const int input_depth = input_shape.Dims(3);
+    const int filter_height = filter_shape.Dims(1);
+    const int filter_width = filter_shape.Dims(2);
+    const int output_height = output_shape.Dims(1);
+    const int output_width = output_shape.Dims(2);
+    const int filter_depth = filter_shape.Dims(3);
+    TFLITE_DCHECK_EQ(output_depth, input_depth * depth_multiplier);
+    TFLITE_DCHECK_EQ(bias_shape.FlatSize(), output_depth);
+
+    int32_t err, i, input_data_format = 0, output_data_format = 0;
+    void *p_scratch;
+    uint8 *p_filter;
+    int filter_depth_padded, filter_size_padded, required_scratch;
+    int input_precision = PREC_ASYM8;
+    int h,c;
+
+    ALLOCATE_XTENSA_NNLIB_SCRATCH_MEM;
+    p_scratch = xtensa_nnlib_scratch_buf;
+
+    required_scratch =  xa_nn_conv2d_depthwise_getsize(
+        input_height,
+        input_width,
+        input_depth,
+        filter_height,
+        filter_width,
+        depth_multiplier,
+        stride_width,
+        stride_height,
+        pad_width,
+        pad_height,
+        output_height,
+        output_width,
+        input_precision,
+        input_data_format);
+
+    if(required_scratch <= 0) {
+      context->ReportError(context, "DepthwiseConvAsym8: xa_nn_conv2d_depthwise_getsize failed");
+      return kTfLiteError;
+    }
+
+    filter_depth_padded = (filter_depth + 3) & (~3);
+    filter_size_padded = filter_height * filter_width * filter_depth_padded;
+    required_scratch += ALIGNED_SIZE(sizeof(uint8_t)* filter_size_padded, 8);
+
+    if(required_scratch > (int)XTENSA_NNLIB_MAX_SCRATCH_SIZE) {
+      context->ReportError(context, "DepthwiseConvAsym8: insufficient scratch memory");
```

Comment:
`TF_LITE_KERNEL_LOG should be used instead of context->ReportError, so log strings can be removed to save space if needed.`